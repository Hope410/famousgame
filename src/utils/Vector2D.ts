import _ from 'lodash';

import { DELIMITER, ERRORS } from '@/config';
import { Tuple2D } from '@/types';

export default class Vector2D {
  x: number;

  y: number;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  static fromKey(key: string) {
    const parts = key.split(DELIMITER).map(Number);

    if (_.some(parts, _.isNaN)) {
      throw new TypeError(ERRORS.INVALID_KEY_PARTS);
    }

    if (parts.length !== 2) {
      throw new TypeError(ERRORS.INVALID_KEY_LENGHT);
    }

    return new Vector2D(...(parts as Tuple2D<number>));
  }

  static fromTuple(tuple: Tuple2D<number>): Vector2D {
    return new Vector2D(...tuple);
  }

  static get top(): Vector2D {
    return new Vector2D(0, -1);
  }

  static get right(): Vector2D {
    return new Vector2D(1, 0);
  }

  static get bottom(): Vector2D {
    return new Vector2D(0, 1);
  }

  static get left(): Vector2D {
    return new Vector2D(-1, 0);
  }

  static get directions() {
    return [
      Vector2D.top,
      Vector2D.right,
      Vector2D.bottom,
      Vector2D.left,
    ] as const;
  }

  asKey(): string {
    return this.asTuple().join(DELIMITER);
  }

  asTuple(): Tuple2D<number> {
    return [this.x, this.y];
  }

  add(vec: Vector2D): Vector2D {
    return new Vector2D(this.x + vec.x, this.y + vec.y);
  }

  subtract(vec: Vector2D): Vector2D {
    return new Vector2D(this.x - vec.x, this.y - vec.y);
  }
}
