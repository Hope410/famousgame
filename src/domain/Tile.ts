import Vector2D from '@/utils/Vector2D';

interface TileOptions {
  position: Vector2D;
  color: string;
}

export default class Tile {
  position: Vector2D;

  color: string;

  constructor(options: TileOptions) {
    this.position = options.position;
    this.color = options.color;
  }

  setColor(color: string) {
    this.color = color;
  }
}
