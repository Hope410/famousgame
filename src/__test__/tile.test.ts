import Tile from '@/domain/Tile';

import Vector2D from '@/utils/Vector2D';

describe('Tile test', () => {
  test('Tile creates without errors', () => {
    const tile = new Tile({
      position: new Vector2D(0, 0),
      color: 'red',
    });
    expect(tile).toMatchSnapshot();
  });

  test('Tile changes color without errors', () => {
    const tile = new Tile({
      position: new Vector2D(0, 0),
      color: 'red',
    });
    expect(tile).toMatchSnapshot();

    tile.setColor('blue');
    expect(tile).toMatchSnapshot();
  });
});
