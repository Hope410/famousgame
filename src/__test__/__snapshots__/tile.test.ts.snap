// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile test Tile changes color without errors 1`] = `
Tile {
  "color": "red",
  "position": Vector2D {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`Tile test Tile changes color without errors 2`] = `
Tile {
  "color": "blue",
  "position": Vector2D {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`Tile test Tile creates without errors 1`] = `
Tile {
  "color": "red",
  "position": Vector2D {
    "x": 0,
    "y": 0,
  },
}
`;
